src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -1256,18 +1267,44 @@ bool ITokenCacheInternal.HasTokensNoLocks()
 
             foreach (MsalIdTokenCacheItem idTokenCacheItem in idTokens)
             {
-                Accessor.DeleteIdToken(idTokenCacheItem);
+                accessor.DeleteIdToken(idTokenCacheItem);
             }
 
             requestContext.Logger.Info($"Deleted {idTokens.Count} ID tokens.");
 
-            var accounts = Accessor.GetAllAccounts(partitionKey);
+            var accounts = accessor.GetAllAccounts(partitionKey);
             accounts.RemoveAll(item => !(item.HomeAccountId.Equals(account.HomeAccountId.Identifier, StringComparison.OrdinalIgnoreCase) &&
                                item.PreferredUsername.Equals(account.Username, StringComparison.OrdinalIgnoreCase)));
 
             foreach (MsalAccountCacheItem accountCacheItem in accounts)
             {
-                Accessor.DeleteAccount(accountCacheItem);
+                accessor.DeleteAccount(accountCacheItem);
+            }
+
+            if (!((ITokenCacheInternal)this).IsAppSubscribedToSerializationEvents())
+            {
+                DateTimeOffset? cacheExpiry = CalculateSuggestedCacheExpiry(accessor, requestContext.Logger);
+                await IdentityCacheWrapper.SetAsync(partitionKey, accessor, cacheExpiry).ConfigureAwait(false);
+            }
+        }
+
+        private async Task<ITokenCacheAccessor> GetOrCreateAccessorAsync(string partitionKey)
```

Comment:
I think the order of precedence of which cache to use should be something like:
- Use user-provided IIdentityCache implementation, or
- Use user-specified legacy cache serialization, or
- Use default in-memory cache based on IIdentityCache.